normandy/base/middleware.py

```diff
@@ -117,3 +122,150 @@ def process_request(self, request):
+            response = self.get_response(request)
+        return response
+
+    def process_request(self, request):
+        header_value = request.META.get('HTTP_AUTHORIZATION')
+        if not header_value:
+            # That's cool. If it's not there, it's not there.
+            # This middleware will not be able to sign in the user
+            # for the duration of this request.
+            return
+        matched = re.match(
+            # Note thet re.match requires it to match from the beginning
+            # otherwise it might match `Something Bearer deadbeef`
+            r'Bearer\s+(.*)',
+            header_value,
+            re.I
+        )
+        if not matched:
+            # The header was there but it didn't have a conformant "Bearer token value".
+            raise PermissionDenied('Invalid Authorization header value')
+
+        access_token = matched.group(1)
+
+        user_profile = self._fetch_oidc_email(access_token)
+        email = user_profile.get('email')
+        if not email:
+            # This would happen if someone has requested an access token
+            # from their OIDC provider *without the 'email' scope*.
+            # For example, if you do:
+            #
+            # new auth0.WebAuth({
+            #   domain: OIDC_DOMAIN,
+            #   clientID: OIDC_CLIENT_ID,
+            #   responseType: 'token id_token',
+            #   scope: 'openid profile',       <-- note lack of 'email'
+            # });
+            #
+            # Ideally the client should review their config. The
+            # access token did work but not sufficient for our needs.
+            raise PermissionDenied("User profile lacks 'email' scope.")
+
+        # Minimal normalization. Also forces all emails to be lower case
+        # which makes the UserModel lookup by email faster.
+        email = email.strip().lower()
+
+        # Awesome! Turn this email into a Django User instance.
+        try:
+            user = self.UserModel.objects.get(email=email)
+        except self.UserModel.DoesNotExist:
+            user = self.UserModel.objects.create(
+                username=email[:150],
+                email=email,
+            )
+
+        # Take this opportunity to suck bits from the user profile to
+        # enrich our current Django user.
+        if (
+            'family_name' in user_profile and 'given_name' in user_profile and
+            (
+                user_profile['family_name'] != user.last_name or
+                user_profile['given_name'] != user.first_name
+            )
+        ):
+            user.first_name = user_profile['given_name'].strip()
+            user.last_name = user_profile['family_name'].strip()
+            user.save()
+
+            # We *could* do something like this too:
+            # if user_profile.get('picture'):
+            #     pf = user.get_profile()
+            #     pf.picture = user_profile['picture']
+            #     pf.save()
```

Comment:
I'd like to either do this or remove the commented code.

Reply:
Agreed.